### PR TITLE
feat: Enable copying of publicly shared question code (if allowed by instructor)

### PR DESCRIPTION
**Pull request by SethPoulsen**
implements parts of #9397

### CODE REVIEW

File: apps/prairielearn/src/pages/instructorCopyTemplateCourseQuestion/instructorCopyTemplateCourseQuestion.ts [hunk 1]
```diff
@@ -39,8 +39,8 @@ router.post(
       }),
     );
 
-    if (result.course.template_course === false) {
-      throw new error.HttpStatusError(400, 'The given course is not a template course.');
+    if (!(result.course.template_course || result.question.shared_publicly_source)) {
+      throw new error.HttpStatusError(400, 'Copying this question is not permitted');
     }
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/instructorCopyTemplateCourseQuestion/instructorCopyTemplateCourseQuestion.ts:44]: Is there a reason to have a separate endpoint for copying from a template course and copying a publicly shared question? Or should I just rename this endpoint to `instructorCopyPublicQuestion`, and use it both for publicly shared questions and template course questions?

Comment (SethPoulsen) [reply apps/prairielearn/src/pages/instructorCopyTemplateCourseQuestion/instructorCopyTemplateCourseQuestion.ts:44]: and what should I do with the URL? its currently `'/pl/course/:course_id(\\d+)/copy_template_course_question'`

Comment (SethPoulsen) [reply apps/prairielearn/src/pages/instructorCopyTemplateCourseQuestion/instructorCopyTemplateCourseQuestion.ts:44]: changed to `copy_public_question`

**Comment by SethPoulsen** [comment]
I am aware that this doesn't actually make a way to mark the `publicly_shared_source` column as true in the database. 

Doing that is going to be a bunch of UI design/mess on the question settings page, I thought it would be nice to do that in a separate PR from this.

File: apps/prairielearn/src/server.js [hunk 2]
```diff
@@ -1869,19 +1869,16 @@ export async function initExpress() {
...
   // pages is that the source question pages are served by chunk servers, but the
   // question transfer machinery relies on access to course repositories on disk,
   // which don't exist on chunk servers
   app.use(
-    '/pl/course/:course_id(\\d+)/copy_template_course_question',
-    (
-      await import(
-        './pages/instructorCopyTemplateCourseQuestion/instructorCopyTemplateCourseQuestion.js'
-      )
-    ).default,
```
Comment (nwalters512) [inline apps/prairielearn/src/server.js:1884]: Removing a route like this is generally not safe; what happens when someone tries to copy a question during a deploy? I agree that we want to rename this route, but we should do it with roughly the same strategy we'd use for any database columns:

- Add the new route; deploy
- Update anything referencing the old route to instead point to the new route; deploy
- Remove the old route

File: apps/prairielearn/src/migrations/20240501185744_questions__shared_publicly_source__add.sql [hunk 3]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE questions
+ADD COLUMN shared_publicly_source BOOLEAN NOT NULL DEFAULT FALSE;
```
Comment (nwalters512) [inline apps/prairielearn/src/migrations/20240501185744_questions__shared_publicly_source__add.sql:2]: Bikeshedding: `shared_publicly_with_source`? `source_shared_publicly`? Something else? The current column name feels strange to me. Let's discuss.

Comment (mwest1066) [reply apps/prairielearn/src/migrations/20240501185744_questions__shared_publicly_source__add.sql:2]: I like `shared_publicly_with_source` to keep the column sorted near the existing `shared_publicly`.

Comment (SethPoulsen) [reply apps/prairielearn/src/migrations/20240501185744_questions__shared_publicly_source__add.sql:2]: I agree `shared_publicly_with_source` is best, thanks for the suggestion @nwalters512 . The alphabetic sorting thing is exactly the reason I didn't do `source_shared_publicly` in the first place.

**Comment by SethPoulsen** [comment]
I created the following two PRs to complete deprecating the old copy endpoint: 

https://github.com/PrairieLearn/PrairieLearn/pull/9918
https://github.com/PrairieLearn/PrairieLearn/pull/9919

File: apps/prairielearn/src/server.js (hunk 2 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/server.js]: done here:

https://github.com/PrairieLearn/PrairieLearn/pull/9918
https://github.com/PrairieLearn/PrairieLearn/pull/9919

File: apps/prairielearn/src/lib/copy-question.ts [hunk 4]
```diff
@@ -19,7 +19,7 @@ const sql = sqldb.loadSqlEquiv(import.meta.url);
 export async function setQuestionCopyTargets(res: Response) {
   // Avoid querying for editable courses if we won't be able to copy this
   // question anyways.
-  if (!res.locals.course.template_course) {
+  if (!(res.locals.course.template_course || res.locals.question.shared_publicly_with_source)) {
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/copy-question.ts:22]: Total nit: IMO the version of this where `!` is distributed is easier to parse, what do you think?

```suggestion
  if (!res.locals.course.template_course && !res.locals.question.shared_publicly_with_source) {
```

Ditto for the similar conditional on the `instructorCopyPublicQuestion` page. Feel free to ignore if you disagree.

**Review by nwalters512 — APPROVED** [review]

**Pull request merged**